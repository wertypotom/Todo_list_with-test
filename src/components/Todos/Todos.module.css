.container {
  background-color: white;

  display: flex;
  flex-direction: column;
  padding: 10px;
  box-shadow: 0px 0px 10px rgba(0, 0, 0, 0.1);

  border-radius: var(--radius);
}

.todo-container {
  max-height: 300px;
  overflow-x: auto;
}

.btn {
  cursor: pointer;
}

.btn:hover {
  color: var(--secondary-color-accent_hover);
}

.footer {
  margin: 8px 0;

  color: var(--secondary-color-accent);
  font-weight: 100;

  display: flex;
  align-items: center;
  justify-content: space-between;
}

.filters {
  font-size: 15px;

  cursor: pointer;
  display: flex;
  align-items: center;
  gap: 15px;
}

.active {
  padding: 3px;

  border: 1px solid var(--accent-color);
  border-radius: var(--radius);
}

::-webkit-scrollbar {
  width: 10px;
}

::-webkit-scrollbar-track {
  background: #f1f1f1;
}

::-webkit-scrollbar-thumb {
  background: #888;
}

::-webkit-scrollbar-thumb:hover {
  background: #555;
}
