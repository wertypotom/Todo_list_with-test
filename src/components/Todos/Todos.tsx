import { useMemo, useState } from 'react';
import { Filters, Todo } from '@/types/Todo.type';
import TodoItem from './TodoItem/TodoItem';
import styles from './Todos.module.css';

type Props = {
  todos: Todo[];
  handleTodoCheck: (todo: Todo) => void;
  handleClearCompletedTodos: () => void;
};

const Todos = ({
  todos,
  handleTodoCheck,
  handleClearCompletedTodos,
}: Props) => {
  const [activeFiter, setActiveFilter] = useState<Filters>('All');

  const todosToRender = useMemo(() => {
    switch (activeFiter) {
      case 'Active':
        return todos.filter((t) => !t.completed);
      case 'Completed':
        return todos.filter((t) => t.completed);
      default:
        return todos;
    }
  }, [activeFiter, todos]);

  const activeTodos = useMemo(() => todos.filter((t) => !t.completed), [todos]);

  return (
    <section className={styles.container}>
      <div className={styles['todo-container']}>
        {todosToRender.map((todo) => (
          <TodoItem
            todo={todo}
            key={todo.id}
            handleTodoCheck={handleTodoCheck}
          />
        ))}
      </div>
      <footer className={styles.footer}>
        <div>{activeTodos.length} items left</div>
        <div className={styles.filters}>
          <div
            onClick={() => setActiveFilter('All')}
            className={`${styles.btn} ${
              activeFiter === 'All' ? styles.active : ''
            }`}
          >
            All
          </div>
          <div
            onClick={() => setActiveFilter('Active')}
            className={`${styles.btn} ${
              activeFiter === 'Active' ? styles.active : ''
            }`}
          >
            Active
          </div>
          <div
            className={`${styles.btn} ${
              activeFiter === 'Completed' ? styles.active : ''
            }`}
            onClick={() => setActiveFilter('Completed')}
          >
            Completed
          </div>
        </div>
        <div
          data-testid='clear-btn'
          className={styles.btn}
          onClick={handleClearCompletedTodos}
        >
          Clear completed
        </div>
      </footer>
    </section>
  );
};

export default Todos;
