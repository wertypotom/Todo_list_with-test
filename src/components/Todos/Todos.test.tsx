import { Todo } from '@/types/Todo.type';
import { render } from '@testing-library/react';
import Todos from './Todos';
import userEvent from '@testing-library/user-event';

const todos: Todo[] = [
  {
    completed: false,
    id: '0',
    title: 'Hello',
  },
  {
    completed: true,
    id: '1',
    title: 'World',
  },
];

describe('Todo Component', () => {
  it('Should render Todo list', () => {
    const { container } = render(
      <Todos
        handleClearCompletedTodos={() => {}}
        handleTodoCheck={() => {}}
        todos={[]}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('Should render Todo list with items', () => {
    const { getByText } = render(
      <Todos
        handleClearCompletedTodos={() => {}}
        handleTodoCheck={() => {}}
        todos={todos}
      />
    );

    const text = getByText(/hello/i);
    expect(text).toBeInTheDocument();
  });

  it('handleClearCompletedTodos should be clickable', async () => {
    const handleClearCompletedTodosMock = vi.fn();

    const { getByTestId } = render(
      <Todos
        handleClearCompletedTodos={handleClearCompletedTodosMock}
        handleTodoCheck={() => {}}
        todos={[]}
      />
    );

    const clearBtn = getByTestId('clear-btn');
    await userEvent.click(clearBtn);
    expect(handleClearCompletedTodosMock).toHaveBeenCalledOnce();
  });

  it('Should correctly dislay remaining Todos', async () => {
    const { getByText } = render(
      <Todos
        handleClearCompletedTodos={() => {}}
        handleTodoCheck={() => {}}
        todos={todos}
      />
    );

    const remainingItems = getByText(/1 items left/i);
    expect(remainingItems).toBeInTheDocument();
  });
});
