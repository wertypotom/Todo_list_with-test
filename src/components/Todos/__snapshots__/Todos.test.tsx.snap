// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Todo Component > Should render Todo list 1`] = `
<div>
  <section
    class="_container_afcc56"
  >
    <div
      class="_todo-container_afcc56"
    />
    <footer
      class="_footer_afcc56"
    >
      <div>
        0
         items left
      </div>
      <div
        class="_filters_afcc56"
      >
        <div
          class="_btn_afcc56 _active_afcc56"
        >
          All
        </div>
        <div
          class="_btn_afcc56 "
        >
          Active
        </div>
        <div
          class="_btn_afcc56 "
        >
          Completed
        </div>
      </div>
      <div
        class="_btn_afcc56"
        data-testid="clear-btn"
      >
        Clear completed
      </div>
    </footer>
  </section>
</div>
`;
